x509-identity-mgmt/js/src/services/CertificateService.js

```diff
@@ -0,0 +1,309 @@
+/**
+ * Service to handle certificates
+ */
+class CertificateService {
+  /**
+   * The dependencies are injected through the constructor
+   */
+  constructor({
+    trustedCAService, certificateModel, ejbcaFacade, tenant, pkiUtils,
+    dnUtils, certValidity, checkPublicKey, queryMaxTimeMS, certMinimumValidityDays,
+    caCertAutoRegistration, logger, errorTemplate,
+  }) {
+    Object.defineProperty(this, 'trustedCAService', { value: trustedCAService });
+    Object.defineProperty(this, 'ejbcaFacade', { value: ejbcaFacade });
+    Object.defineProperty(this, 'tenant', { value: tenant });
+    Object.defineProperty(this, 'pkiUtils', { value: pkiUtils });
+    Object.defineProperty(this, 'dnUtils', { value: dnUtils });
+    Object.defineProperty(this, 'certValidity', { value: certValidity });
+    Object.defineProperty(this, 'checkPublicKey', { value: checkPublicKey });
+    Object.defineProperty(this, 'queryMaxTimeMS', { value: queryMaxTimeMS });
+    Object.defineProperty(this, 'certMinimumValidityDays', { value: certMinimumValidityDays });
+    Object.defineProperty(this, 'caCertAutoRegistration', { value: caCertAutoRegistration });
+    Object.defineProperty(this, 'logger', { value: logger });
+    Object.defineProperty(this, 'CertificateModel', { value: certificateModel.model });
+    Object.defineProperty(this, 'error', { value: errorTemplate });
+  }
+
+  /**
+   * Generates an x509 certificate based on a CSR
+   *
+   * @param {object} Object with the CSR and data whom the certificate is to be associated.
+   *
+   * @returns an object containing the certificate in PEM format and its fingerprint.
+   */
+  async generateCertificate({ csr: csrPem, belongsTo }) {
+    const csr = this.pkiUtils.parseCSR(csrPem);
+
+    if (this.checkPublicKey) {
+      this.pkiUtils.checkPublicKey(csr.subjectPublicKeyInfo);
+    }
+
+    const subjectDN = this.dnUtils.from(csr.subject, true)
+      .verify()
+      .cnamePrefix(this.tenant)
+      .stringify();
+
+    const certificatePem = await this.ejbcaFacade.generateCertificate(
+      subjectDN, this.certValidity, csrPem,
+    );
+
+    const cert = this.pkiUtils.parseCert(certificatePem);
+    const certificateFingerprint = this.pkiUtils.getFingerprint(certificatePem);
+
+    const model = new this.CertificateModel({
+      fingerprint: certificateFingerprint,
+      pem: certificatePem,
+      belongsTo,
```

Comment:
For certificates signed by dojot, we could force `cnames` to be equal to `belongsTo`. What do you think?

Reply:
I think that all CSR `Subject DN` fields could be ignored and dojot could generate a UUIDv4 as the value of de `cname` field in the issued certificate.

Reply:
I agree with you!